ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/PPLTool.java

```diff
@@ -0,0 +1,323 @@
+            if (INSTANCE != null) {
+                return INSTANCE;
+            }
+            synchronized (PPLTool.class) {
+                if (INSTANCE != null) {
+                    return INSTANCE;
+                }
+                INSTANCE = new PPLTool.Factory();
+                return INSTANCE;
+            }
+        }
+
+        public void init(Client client) {
+            this.client = client;
+        }
+
+        @Override
+        public PPLTool create(Map<String, Object> map) {
+            return new PPLTool(client, (String)map.get("model_id"));
+        }
+
+        @Override
+        public String getDefaultDescription() {
+            return DEFAULT_DESCRIPTION;
+        }
+    }
+
+    private SearchRequest buildSearchRequest(String indexName)
+    {
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+        searchSourceBuilder
+                .size(1)
+                .query(new MatchAllQueryBuilder());
+        //  client;
+        SearchRequest request = new SearchRequest(new String[]{indexName}, searchSourceBuilder);
+        return request;
+    }
+
+    private GetMappingsRequest buildGetMappingRequest(String indexName)
+    {
+        String [] indices = new String[] {indexName};
+        GetMappingsRequest getMappingsRequest = new GetMappingsRequest();
+        getMappingsRequest.indices(indices);
+        return getMappingsRequest;
+    }
+
+    private String constructTableInfo(SearchHit [] searchHits, Map<String, MappingMetadata> mappings, String indexName)
+    {
+        MappingMetadata mappingMetadata = mappings.get(indexName);
+        Map<String, Object> mappingSource = (Map<String, Object>) mappingMetadata.getSourceAsMap().get("properties");
+        Map<String, String> fieldsToType = new HashMap<>();
+        extractNamesTypes(mappingSource, fieldsToType, "");
+        StringJoiner tableInfoJoiner = new StringJoiner("\n");
+
+        if (searchHits.length > 0) {
+            SearchHit hit = searchHits[0];
+            Map<String, Object> sampleSource = hit.getSourceAsMap();
+            Map<String, String> fieldsToSample = new HashMap<>();
+            for (String key : fieldsToType.keySet()) {
+                fieldsToSample.put(key, "");
+            }
+            extractSamples(sampleSource, fieldsToSample, "");
+
+
+            for (String key : fieldsToType.keySet()) {
+                String line = "- " + key + ": " + fieldsToType.get(key) + " (" + fieldsToSample.get(key) + ")";
+                tableInfoJoiner.add(line);
+            }
+        }
+        else
+        {
+            for (String key : fieldsToType.keySet()) {
+                String line = "- " + key + ": " + fieldsToType.get(key);
+                tableInfoJoiner.add(line);
+            }
+        }
+
+        String tableInfo = tableInfoJoiner.toString();
+        return tableInfo;
+    }
+
+    private String constructPrompt(String tableInfo, String question, String indexName)
```

Comment:
Prompt should not be hard coded in tool but rather be passed as tool spec. This provides flexibility to modify prompt and allow the tool to operate with different remote models.

Reply:
Already add a parameter called contextPrompt. But PPL tool needs a prompt template containing position for question, index and table information. We need to have a detailed description on that so our customer can register it correctly.